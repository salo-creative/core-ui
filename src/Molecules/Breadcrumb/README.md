# Breadcrumb

This is a component for displaying navigational breadcrumbs

---

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Divider from '@salo/core-ui/Molecules/Breadcrumb';
```

Implement as follows

```javascript
<Breadcrumb
  margin='0 0 2rem'
  trail={ [
    { label: 'Home', link: 'https://www.google.co.uk'},
    { label: 'Previous page', link: '/previous-page' },
    { label: 'Current page' }
  ] }
>
```
If you pass a link url to an item in the breadcrumb it will by default render a React Link component for internal routing. If you need to use an external url in your breadcrumb simply include the protocol (don't do this bad UX!)

**For full prop types and usage see storybook info/knobs**