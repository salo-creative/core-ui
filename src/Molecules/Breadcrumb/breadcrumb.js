import React from 'react';
import PropTypes from 'prop-types';

// COMPONENTS & STYLES
import Item from './breadcrumb.item';
import { Wrapper } from './breadcrumb.styles';

const Breadcrumb = (props) => {
  const { margin, trail } = props;

  return (
    <Wrapper margin={ margin }>
      { trail.map((item, i) => {
        const lastItem = i === (trail.length - 1);
        return (
          <Item
            key={ item.label }
            lastItem={ lastItem }
            { ...item }
          />
        );
      }) }
    </Wrapper>
  );
};

Breadcrumb.defaultProps = {
  margin: '0 0 2rem',
  trail: []
};

Breadcrumb.propTypes = {
  margin: PropTypes.string,
  trail: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    link: PropTypes.string
  }))
};

export default Breadcrumb;