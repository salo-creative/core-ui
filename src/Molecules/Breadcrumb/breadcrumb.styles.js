import styled from 'styled-components';
import { darken } from 'polished';

export const Wrapper = styled.div`
  width: 100%;
  display: flex;
  height: 5rem;
  margin: ${ ({ margin }) => margin };
  background: ${ ({ theme }) => theme.grey };
  border-radius: 0.4rem;
  padding: 0 1rem;
`;

export const ItemWrapper = styled.span`
  display: flex;
  align-items: center;
  height: 100%;
  font-size: 1.4rem;
  a {
    height: 100%;
    display: flex;
    align-items: center;
    text-decoration: none;
    color: ${ ({ theme }) => darken(0.2, theme.blue) };
    transition: color 0.3s linear;
    &:hover {
      color: ${ ({ theme }) => theme.blue };
    }
  }
`;

export const Divider = styled.span`
  color: ${ ({ theme }) => theme.darkGrey };
  display: flex;
  align-items: center;
  padding: 0 1rem;
  height: 100%;
`;