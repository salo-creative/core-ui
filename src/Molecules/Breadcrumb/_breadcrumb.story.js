import React from 'react';
import { storiesOf } from '@storybook/react';

// load tests
import { withTests } from '@storybook/addon-jest';
import { withKnobs, text, object } from '@storybook/addon-knobs';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Breadcrumb } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Molecules | Breadcrumb', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['breadcrumb'] });

stories.add(
  'Basic',
  (() => {
    const margin = text('Margin', '0 0 2rem');
    const trail = object('Trail', [
      { label: 'Home', link: 'https://www.google.co.uk' },
      { label: 'Previous page', link: '/previous-page' },
      { label: 'Current page' }
    ]);
    return (
      <Breadcrumb
        margin={ margin }
        trail={ trail }
      />
    );
  }), { notes: README }
);