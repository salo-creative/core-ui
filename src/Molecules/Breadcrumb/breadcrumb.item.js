import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

// COMPONENTS & STYLES
import { ItemWrapper, Divider } from './breadcrumb.styles';

const Item = (props) => {
  const { label, lastItem, link } = props;

  // Render the item as link or text
  const renderItem = () => {
    if (link && link.includes('://')) {
      return <a href={ link } title={ label }>{ label }</a>;
    }
    if (link) {
      return <Link to={ link }>{ label }</Link>;
    }
    return label;
  };

  return (
    <ItemWrapper>
      { renderItem() }
      { !lastItem && (
        <Divider>/</Divider>
      ) }
    </ItemWrapper>
  );
};

Item.defaultProps = {
  lastItem: false,
  link: ''
};

Item.propTypes = {
  label: PropTypes.string.isRequired,
  lastItem: PropTypes.bool,
  link: PropTypes.string
};

export default Item;