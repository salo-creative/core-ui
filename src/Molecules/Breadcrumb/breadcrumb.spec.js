import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Breadcrumb from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Breadcrumb />
  );
  expect(container.firstChild).toMatchSnapshot();
});