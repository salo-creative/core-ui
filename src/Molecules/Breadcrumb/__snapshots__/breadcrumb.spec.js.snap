// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 5rem;
  margin: 0 0 2rem;
  background: #e9ecef;
  border-radius: 0.4rem;
  padding: 0 1rem;
}

<div
  class="c0"
/>
`;
